### PR TITLE
Add support for Basic.Return

### LINKED ISSUE #140: basic.publish mandatory bit broken
It seems that `mandatory` is not really implemented. `mandatory=False` works, but any attemp with `True` results in:

```
[2017-05-16 10:10:38,767] ERROR aioamqp.protocol error on dispatch
    Traceback (most recent call last):
      File "/lib/python3.5/site-packages/aioamqp/protocol.py", line 333, in run
        yield from self.dispatch_frame()
      File "/lib/python3.5/site-packages/aioamqp/protocol.py", line 288, in dispatch_frame
        yield from channel.dispatch_frame(frame)
      File "/lib/python3.5/site-packages/aioamqp/channel.py", line 106, in dispatch_frame
        raise NotImplementedError("Frame (%s, %s) is not implemented" % (frame.class_id, frame.method_id))
    NotImplementedError: Frame (None, None) is not implemented
```

It would be nicer to have such constraints documented in method docstrings or to remove unimplemented parameters from public methods.

### PR DESCRIPTION
**Pull request by nhumrich**
Fixes #140

### CODE REVIEW

**Comment by nhumrich** [comment]
Bumb. Any word on this? Can this get merged, or can we at least have a discussion about it?

**Comment by Relrin** [comment]
@dzen @RemiCardona Any plans to merge this PR ?

**Comment by dzen** [comment]
@nhumrich Hello.
Sorry for the very late aswer. 
We have a few comments on this PR but we're open to this !

File: aioamqp/channel.py [hunk 1]
```diff
@@ -778,6 +780,32 @@ def basic_recover_ok(self, frame):
...
+        buffer = io.BytesIO()
+        while buffer.tell() < content_header_frame.body_size:
+            content_body_frame = yield from self.protocol.get_frame()
+            buffer.write(content_body_frame.payload)
+
+        body = buffer.getvalue()
+        envelope = ReturnEnvelope(reply_code, reply_text,
+                                  exchange_name, routing_key)
+        properties = content_header_frame.properties
+        callback = self.return_callback
```
Comment (dzen) [inline aioamqp/channel.py]: This should be added as an optional keyword parameter to the `queue_declare` coroutine

Comment (nhumrich) [reply aioamqp/channel.py]: Should it be part of queue declare? I am not so sure. A queue isnt related to a retun at all. You have get a return packet without ever declaring a queue. Not sure they are related.

**Comment by nhumrich** [comment]
Added test, and rebased onto master

File: aioamqp/channel.py (hunk 1 above)
Comment (dzen) [reply aioamqp/channel.py]: Oh, got it, I just checked the doc.

To be honest, I think it should be declared at channel creation time.
Imho, an addition to the `Channel`'s `__init__` function with a default to None should work shouldn't it ?

So, each time we publish a message through a channel, we could set the function when creating it, to receive failed messages.

**Comment by nhumrich** [comment]
Sounds good to me. Updated.

File: aioamqp/tests/test_publish.py [hunk 2]
```diff
@@ -51,3 +52,28 @@ def test_confirmed_publish(self):
...
+    @testing.coroutine
+    def test_return_from_publish(self):
+        called = False
+
+        @asyncio.coroutine
+        def callback(channel, body, envelope, properties):
+            nonlocal called
+            called = True
+
+        self.channel.return_callback = callback
```
Comment (dzen) [inline aioamqp/tests/test_publish.py]: you should create a new channel, with the parameter in the init() or in the factory `self.amqp.create_channel(callback=callback)` IIRC

File: examples/send_with_return.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+          'Return code: {}\n'
+          'Return message: {}\n'
+          'exchange: {}'.format(envelope.routing_key, envelope.reply_code,
+                                envelope.reply_text, envelope.exchange_name))
+
+
+@asyncio.coroutine
+def send():
+    transport, protocol = yield from aioamqp.connect()
+    channel = yield from protocol.channel()
```
Comment (dzen) [inline examples/send_with_return.py]: don't miss the callback here :)

**Comment by nhumrich** [comment]
Updated. Fixed test and example.

**Comment by dzen** [comment]
thanks ! will merge soon !

**Comment by dzen** [comment]
Thanks for your contribution. 
I added your name in the authors file.

**Pull request merged**